nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/RestLookupService.java

```diff
@@ -0,0 +1,390 @@
+                + "for authentication.")
+        .required(false)
+        .defaultValue("false")
+        .allowableValues("true", "false")
+        .build();
+
+    private static final ProxySpec[] PROXY_SPECS = {ProxySpec.HTTP_AUTH, ProxySpec.SOCKS};
+    public static final PropertyDescriptor PROXY_CONFIGURATION_SERVICE
+            = ProxyConfiguration.createProxyConfigPropertyDescriptor(true, PROXY_SPECS);
+
+    static final String ENDPOINT_KEY = "endpoint";
+    static final String MIME_TYPE_KEY = "mime.type";
+    static final String BODY_KEY = "request.body";
+    static final String METHOD_KEY = "request.method";
+
+    static final List<PropertyDescriptor> DESCRIPTORS;
+
+    static {
+        DESCRIPTORS = Collections.unmodifiableList(Arrays.asList(
+            RECORD_READER,
+            RECORD_PATH,
+            SSL_CONTEXT_SERVICE,
+            PROXY_CONFIGURATION_SERVICE,
+            PROP_BASIC_AUTH_USERNAME,
+            PROP_BASIC_AUTH_PASSWORD,
+            PROP_DIGEST_AUTH
+        ));
+    }
+
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return DESCRIPTORS;
+    }
+
+    private ProxyConfigurationService proxyConfigurationService;
+    private RecordReaderFactory readerFactory;
+    private RecordPath recordPath;
+    private OkHttpClient client;
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        proxyConfigurationService = context.getProperty(PROXY_CONFIGURATION_SERVICE)
+                .asControllerService(ProxyConfigurationService.class);
+
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+
+        setAuthenticator(builder, context);
+
+        if (proxyConfigurationService != null) {
+            setProxy(builder);
+        }
+
+        final SSLContextService sslService = context.getProperty(SSL_CONTEXT_SERVICE).asControllerService(SSLContextService.class);
+        final SSLContext sslContext = sslService == null ? null : sslService.createSSLContext(SSLContextService.ClientAuth.WANT);
+        if (sslService != null) {
+            builder.sslSocketFactory(sslContext.getSocketFactory());
+        }
+
+        client = builder.build();
+
+        String path = context.getProperty(RECORD_PATH).isSet() ? context.getProperty(RECORD_PATH).getValue() : null;
+        if (!StringUtils.isBlank(path)) {
+            recordPath = RecordPath.compile(path);
+        }
+
+        getHeaders(context);
+    }
+
+    private Map<String, String> headers;
+    private void getHeaders(ConfigurationContext context) {
+        headers = new HashMap<>();
+        for (PropertyDescriptor descriptor : context.getProperties().keySet()) {
+            if (descriptor.getName().startsWith("header.")) {
```

Comment:
If dynamic properties are used for specifying headers to use, why not just let the property name be the name of the header? I.e., why require a "header." prefix? We should also document how dynamic properties are used via the @DynamicProperties annotation.